torch/distributed/_composable/replicate.py

```diff
@@ -0,0 +1,65 @@
+import torch
+import torch.nn as nn
+
+from typing import List, Tuple
+
+
+class DistributedState:
+    ...
+
+
+class ReplicateState(DistributedState):
+    def __init__(self) -> None:
+        self.modules: List[nn.Module] = []
+        self.parameters: List[nn.Parameter] = []
+        self.has_initialized: bool = False
+
+    def add_modules(self, *modules: nn.Module) -> None:
+        for module in modules:
+            self.modules.append(module)
+            module._distributed_state = self  # type: ignore[assignment]
+            module.register_forward_pre_hook(self.forward_pre_hook)
+            module.register_forward_hook(self.forward_post_hook)
+
+    def _recursive_add_params(self, module: nn.Module) -> None:
+        ...
+
+    def init_helper(self):
+        self.has_initialized = True
+        for module in self.modules:
+            self._recursive_add_params(module)
+        # broadcast parameters, create Reducer
+        ...
+
+    def forward_pre_hook(
+        self, module: nn.Module, input: Tuple[torch.Tensor]
+    ) -> None:
+        if not self.has_initialized:
```

Comment:
Do you mean hook on `self.modules` only, or recursively on all sub-modules of `self.modules`?

My current idea is former, `self.modules` only. Otherwise it may hurt performance. But it may change when we come to more details in implementation.